Sprint -3

kulinichrs opened the pull request.

kulinichrs committed:
Oop Done

kulinichrs committed:
Oop: removed hashcode

kulinichrs committed:
IO - done

kulinichrs committed:
IO - Better solution with USE()

kulinichrs committed:
NIO - Done

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/kotlin_school_2/job/PR_BUILDER/186/

SberKotlinSchool commented on oop/src/main/kotlin/ru/sber/oop/DataClasses.kt:
```diff
@@ -2,10 +2,27 @@ package ru.sber.oop
     lateinit var city: String
+
+    override fun equals(other: Any?): Boolean {
+        return if (other is User) {
```
сперва надо сравнивать ссылки(аргументы) через ===

SberKotlinSchool commented on oop/src/main/kotlin/ru/sber/oop/DataClasses.kt:
```diff
@@ -2,10 +2,27 @@ package ru.sber.oop
-    //TODO: user3 = ...
+    val user2 = user1.copy(name = "Gustav")
+    user1.city = "Omsk"
+    val user3 = user1.copy()
```
тут можно применить apply

SberKotlinSchool commented on oop/src/main/kotlin/ru/sber/oop/Inheritance.kt:
```diff
@@ -1,13 +1,21 @@
 
+    // 1: Переопределил load()  и в качестве придумывания строку переопределил ещё и description
+    override fun description() = "Town Square: $name"
+    final override fun load(): String  = "This is a ${this.description()} "
```
в задании написано что функцию load нельзя переопределять в наследниках. но в котлине все классы по умолчанию final, следовательно класс TownSquare должен быть open, чтобы final на функции load имело смысл

kulinichrs replied:
Получается. чтобы выполнить этот пункт можно было ничего не делать ?

kulinichrs replied:
Поправил. Сделал TownSquare open

SberKotlinSchool commented on oop/src/main/kotlin/ru/sber/oop/Polymorphism.kt:
```diff
@@ -1,13 +1,61 @@
+val Room.monster: Goblin
+    get() = Goblin()
 
+fun Monster.getSalutation(): String {
```
тут можно было использовать короткую запись 
 fun Monster.getSalutation(): = "Hi, im $name"

kulinichrs committed:
QA - DONE

Commit:
Merge pull request #1 from powernano/dev

QA - DONE

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/kotlin_school_2/job/PR_BUILDER/190/

kulinichrs committed:
OOP Fixed:
+ short form for getSalutation()
+ TownSquare changed to open
+ added apply, but its not important
+ links in equal comparede with ===

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/kotlin_school_2/job/PR_BUILDER/191/

kulinichrs commented on qa/src/main/kotlin/ru/sber/qa/HrDepartment.kt:
```diff
@@ -10,8 +10,12 @@ import java.util.LinkedList
+    // пробовал создать локально свой Linked List и подсовывать его примерно вот так:
+    // every { hrDepartment getProperty "incomeBox" } answers { localIn }
+    // Но не пошло
+    public val incomeBox: LinkedList<CertificateRequest> = LinkedList()
```
@rmasgutov, Привет, тут написал коммент прям в коде. Можешь подсказать как тут надо было правильно сделать, чтобы не меняя область видимости или подменить ссылки на списки или просто их получить.

kulinichrs commented on qa/src/test/kotlin/ru/sber/qa/HrDepartmentTest.kt:
```diff
@@ -0,0 +1,160 @@
+        hrDepartment.clock = Clock.fixed(
+            LocalDate.parse("2022-10-" + String.format("%02d", dayOfWeek.getValue() + 2))
+                .atStartOfDay(ZoneId.of("Europe/Paris"))
+                .toInstant(),
```
Уже нашёл правильное решение, но не стал переделывать
`mockkStatic(LocalDateTime::class)`
`every { LocalDateTime.now(any<Clock>()).dayOfWeek } returns dayOfWeek`

kulinichrs reviewed (commented):
Как мне теперь понять какой % покрытия тестами получился? Всё ли я сделал?
